packages/docusaurus-theme-classic/src/theme/TOC/index.tsx

```diff
@@ -42,7 +49,11 @@ export function TOCHeadings({
             // eslint-disable-next-line react/no-danger
             dangerouslySetInnerHTML={{__html: heading.value}}
           />
-          <TOCHeadings isChild toc={heading.children} />
+          <TOCHeadings
+            isChild
+            toc={heading.children}
+            depth={headingLevel + 1}
```

Comment:
I was thinking of this API:

```ts
<TOC maxHeadingLevel={3} />
<TOCCollapsible maxHeadingLevel={3} />
<TOCInline maxHeadingLevel={3} />
```

And then for each level of `TOCHeadings` decrease `depth` by 1:

```ts
  if (depth === 0) {
    return null;
  }
  // ...
        <TOCHeadings
            isChild
            toc={heading.children}
            depth={headingLevel - 1}
  // ...
```

This way you don't need to hook into themeConfig in every `TOCHeadings` and when you use `TOCInline` as a separate component (which is documented usage) you can also customize its depth

Reply:
Addressed in 213289e

Reply:
looks like this has been reverted right? I liked the idea of recursive `depth - 1`

Reply:
> e.g. If someone skips from h2 to h4 and maxHeadingLevel is h3, the current alg will still recurse twice and pick up the h4 in the TOC.

I see.

If the user uses h2 + h4, let's not display h4 in the TOC by default, because it's not retro-compatible with the current behavior.